File: src/TinRoll.Test/Repository/QuestionRepositoryTests.cs
using Microsoft.EntityFrameworkCore;
using Moq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TinRoll.Data;
using TinRoll.Data.Entities;
using TinRoll.Data.Repository;
using TinRoll.Data.Repository.Interface;
using Xunit;

namespace TinRoll.Test.Repository
{
    public class QuestionRepositoryTests : RepositoryTest
    {

        [Fact]
        public async Task Test_Create_Question()
        {

            var newQuestion = new Question
            {
                Id = 1,
                Title = "Unit Test Question",
                UserId = 1,
                Text = "Question Text"
            };

            var mockBaseRepo = new Mock<IBaseRepository<Question>>();

            mockBaseRepo.Setup(m => m.CreateAsync(It.IsAny<Question>()))
                .ReturnsAsync(newQuestion);

            var questionRepo = new QuestionRepository(mockBaseRepo.Object);
            var dbQuestion = await questionRepo.CreateQuestionAsync(newQuestion);

            Assert.NotNull(dbQuestion);
            Assert.Equal(1, dbQuestion.Id);
        }

        [Fact]
        public async Task Test_Get_Question()
        {

            var questionToGet = new Question
            {
                Title = "Unit Test Question",
                UserId = 1,
                Text = "Question Text"
            };

            var mockBaseRepo = new Mock<IBaseRepository<Question>>();

            mockBaseRepo.Setup(m => m.GetAsync(It.IsAny<int>()))
                .ReturnsAsync(questionToGet);

            
            var questionRepo = new QuestionRepository(mockBaseRepo.Object);
            var dbQuestion = await questionRepo.GetQuestionAsync(questionToGet.Id);

            Assert.NotNull(dbQuestion);
            Assert.Equal(questionToGet.Id, dbQuestion.Id);
        }

        [Fact]
        public async Task Test_Get_Questions()
        {
            var questionToGet = new Question
            {
                Title = "Unit Test Question",
                UserId = 1,
                Text = "Question Text"
            };

            var questionToGet2 = new Question
            {
                Title = "Uni Test Question 2",
                UserId = 1,
                Text = "Question Text 2"
            };
            var questions = new List<Question>() { questionToGet, questionToGet2 };

            var mockBaseRepo = new Mock<IBaseRepository<Question>>();

            mockBaseRepo.Setup(m => m.GetAsync(null, null, null))
                .ReturnsAsync(questions);

            var questionRepo = new QuestionRepository(mockBaseRepo.Object);
            var dbQuestions = await questionRepo.GetQuestionsAsync();
            
            Assert.NotNull(dbQuestions);
            Assert.Equal(2, dbQuestions.Count());
        }
    }
}

File: src/TinRoll.Test/Data/Repository/QuestionRepositoryTests.fs
module Data.Repository.QuestionRepositoryTests

open TinRoll.Data
open TinRoll.Data.Repository
open System
open Xunit
open NFluent

let GetIQuestionRepo dbName =
    let options = RepositoryTests.BuildInMemoryDatabase dbName
    let context = new TinRollContext(options)
    new QuestionRepository(context) :> IQuestionRepository


let GetTestQuestion () =
    new Question(0, "Test Title", "Test Text", 0, DateTime.Now, DateTime.Now)

[<Fact>]
let ``Test Create Question`` () =

    let question = GetTestQuestion()
    let questionRepo = GetIQuestionRepo "Test_Create_Question"
    let createdQuestionId = Async.RunSynchronously (questionRepo.CreateQuestionAsync(question))

    Assert.NotNull(1, createdQuestionId)

[<Fact>]
let ``Test Get Question`` () =
    let question = GetTestQuestion()
    let questionRepo = GetIQuestionRepo "Test_Get_Question"
    let createdQuestionId = Async.RunSynchronously (questionRepo.CreateQuestionAsync(question))

    let searchedQuestion = Async.RunSynchronously (questionRepo.GetQuestionAsync createdQuestionId)

    Check.That(Option.isSome searchedQuestion) |> ignore

[<Fact>]
let ``Test Get Questions`` () =
    let firstQuestion = GetTestQuestion()
    let secondQuestion = GetTestQuestion()
    let questionRepo = GetIQuestionRepo "Test_Get_Questions"

    Async.RunSynchronously(questionRepo.CreateQuestionAsync firstQuestion) |> ignore
    Async.RunSynchronously(questionRepo.CreateQuestionAsync secondQuestion) |> ignore

    let questions = Async.RunSynchronously(questionRepo.GetQuestionsAsync())

    Assert.Equal(2, questions.Length)

[<Fact>]
let ``Test Get Question Fail`` () =
    let questionRepo = GetIQuestionRepo "Test_Get_Question_Fail"
    
    let searchedQuestion = Async.RunSynchronously(questionRepo.GetQuestionAsync 23)

    Check.That(Option.isNone searchedQuestion) |> ignore